core/include/traccc/utils/particle.hpp

```diff
@@ -0,0 +1,98 @@
+/** TRACCC library, part of the ACTS project (R&D line)
+ *
+ * (c) 2024 CERN for the benefit of the ACTS project
+ *
+ * Mozilla Public License Version 2.0
+ */
+
+#pragma once
+
+// traccc include(s).
+#include "traccc/definitions/primitives.hpp"
+#include "traccc/definitions/qualifiers.hpp"
+#include "traccc/edm/track_parameters.hpp"
+
+// detray include(s).
+#include "detray/definitions/pdg_particle.hpp"
+
+// System include(s).
+#include <stdexcept>
+
+namespace traccc {
+
+namespace detail {
+
+template <typename scalar_t>
+TRACCC_HOST_DEVICE inline detray::pdg_particle<scalar_t>
+particle_from_pdg_number(const std::int32_t pdg_num) {
+
+    if (pdg_num == 11) {
+        return detray::electron<scalar_t>();
+    } else if (pdg_num == -11) {
+        return detray::positron<scalar_t>();
+    } else if (pdg_num == 13) {
+        return detray::muon<scalar_t>();
+    } else if (pdg_num == -13) {
+        return detray::antimuon<scalar_t>();
+    } else if (pdg_num == 211) {
+        return detray::pion_plus<scalar_t>();
+    } else if (pdg_num == -211) {
+        return detray::pion_minus<scalar_t>();
+    }
+
+    return detray::muon<scalar_t>();
+}
+
+// Apply the charge operator to return the antimatter
+template <typename scalar_t>
+TRACCC_HOST_DEVICE inline detray::pdg_particle<scalar_t> charge_operator(
+    const detray::pdg_particle<scalar_t>& ptc) {
+
+    const auto pdg_num = ptc.pdg_num();
+
+    if (pdg_num == 11) {
```

Comment:
Here is a stupid question: Apart from the nicer naming, can't we just multiply the charge with the sign of the pdg number and avoid the branching?

Reply:
Isn't it rather problematic for majorana particle like gamma

Reply:
Aren't those positive sign numbers? So nothing would change?

Reply:
Also can the charge in out core right now actually be zero? Would that not kill ```qop```?

Reply:
Can you just explain with an example. multiplying the charge with the sign of pdg number sounds confusing to me.

Reply:
Ah, never mind, on the pion the charge already does not correspond to the pdg sign anymore. But can't we then check for neutral particles (which I am not sure the code can handle anyway?) an otherwise just flip the charge?

Reply:
If it just stays as electron with positve charge, it will go through electron bremss instead of positron bremss which does not make sense to me

Reply:
Ah, Ok, makes sense. But I thought on a code level ```electron``` and ```positron``` are not different types, they just have the sign of the pdg number and charge flipped, or is this where I am missing the point? So, I was hoping we could get rid of all this branching here and just flip the respective signs? But maybe not

Reply:
I don't know, e.g.:
```
if (ptc.charge() != 0) {
  return detray::pdg_particle<scalar_t>{-ptc.pdg_num(), ptc.mass(), -ptc.charge()};
} else {
  return ptc;
}
```
but then again, I probably just don't see the full picture here

Reply:
This makes sense but are you sure that there is not any possibility that multiplying negative sign will cause a rounding error? I need to make sure that `(detray::pdg_particle(-electron.pdg_num(), electron.mass(), -electron.charge() == positron())`

Reply:
I was under the impression that this should just flip the signbit

Reply:
OK there is not problem in your suggestion. But let me keep this at the moment. One single mistake in detray pdg definition can make things go south. Probably I will migrate this function to detray and  improve the detray test to make sure everything is OK.

Also don't worry about the branching. As mentioned earlier, the hypothesis is algorithm-wide and it is not likely tracks will be set to totally different types of particle

Reply:
Well, if we are calling this for every CKF step/material interaction etc, the code still has to do some branch checking on the GPU, especially if the default (muon) is not the first hit.

If you are worried just about the equality operator failing, is it not enough to compare the pdg numbers (the material interaction seems to be doing just that actually). These are integers, no rounding will happen there

Reply:
checking if integer is 11 or 13 does not take long time.. this is something that we need to seriously worry about

Reply:
No, the check does not take time and for the CPU this will not be a problem. But even in the examples where we randomize the charge, the warp could very well hit different branches in this code, or not? We need to see at some point what this compiles into, I guess

Reply:
Yeah but that will account for 0.000001 % performance degradation of entire chain probably.

Let me come back to this issue after I make a safe guard in detray OK?

Reply:
OK, sure! Also, if somebody has a particle hypothesis that is not part of this list, say gamma which you mentioned earlier, will their particle just be switched to a muon silently?